File: CodeForces/398/garland.py
import sys
sys.setrecursionlimit(3000000)

def find_temp(node):
    total_temps[node] = temps[node]
    for child in tree[node]:
        find_temp(child)
        total_temps[node] += total_temps[child] 

    if total_temps[node] == x:
        x_list[node] = True
    if total_temps[node] == 2*x:
        twox_list[node] = True

def find(node):
    two_x_found= 0
    where_found = []
    for child in tree[node]:
        x_present, where = find(child)
        if x_present:
            two_x_found += 1
            where_found.append(where)
        if two_x_found == 2:
            print where_found[0], where_found[1]
            exit(0)
        if twox_list[node] and x_present:
            print node, where
            exit(0)
    if x_list[node]:
        two_x_found += 1
        where_found.append(node)
    if two_x_found > 0:
        return True, where_found[0]
    return False, []

n = int(raw_input())
tree = [[] for i in xrange(n+1)]
temps = [0 for i in xrange(n+1)]
total_temps = [0 for i in xrange(n+1)]
total_temp = 0
for i in xrange(n):
    a, t = map(int, raw_input().split())
    if a == 0:
        root = i+1
    if a != 0:
        tree[a].append(i+1)
    temps[i+1] = t
    total_temp += t

x = total_temp / 3
x_list = [False for i in xrange(n+1)]
twox_list = [False for i in xrange(n+1)]
if total_temp % 3 == 0:
    find_temp(root)
    find(root)
print -1

File: CodeForces/398/garland.cpp
import java.util.*;
import java.io.*;

public class garland {

    private static int total_temps[];
    private static ArrayList<Integer> tree[];
    private static int x;
    private static boolean x_present;
    private static int where;
    public static void main(String args[]) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(reader.readLine());
        tree = new ArrayList[n+1];
        for (int i = 0; i < n+1;i++) {
            tree[i] = new ArrayList<Integer>();
        }
        total_temps = new int[n+1];
        int total_temp = 0;
        int root = 1;
        for (int i =0; i < n; i++) {
            StringTokenizer tokenizer = new StringTokenizer(reader.readLine(), " ");
            int a = Integer.parseInt(tokenizer.nextToken());
            int t = Integer.parseInt(tokenizer.nextToken());
            if (a == 0)
                root = i+1;
            tree[a].add(i+1);
            total_temps[i+1] = t;
            total_temp += t;

        } 
        x = total_temp / 3;
        if (total_temp % 3 == 0) {
            find(root);
        }
        System.out.println(-1);
    }


    public static void find(int node) {
        int two_x_found= 0;
        int where_found[] = new int[2];
        for (int child : tree[node]) {
            find(child);
            total_temps[node] += total_temps[child];
            if (x_present && two_x_found < 2) {
                where_found[two_x_found] = where;
                two_x_found += 1;
            }
            
        }
        if (two_x_found == 2) {
            System.out.println(where_found[0] + " " + where_found[1]);
            System.exit(0);
        }
        if (total_temps[node] == 2*x && x_present) {
            System.out.println(node + " " + where);
            System.exit(0);
        }
            
        if (total_temps[node] == x) {
            where_found[two_x_found] = node;
            two_x_found += 1;
        }

        if (two_x_found > 0) {
            x_present = true;
            where = where_found[0];
        }
        else
            x_present = false;
    }

}